### PR TITLE
[Improvement] ShuffleBufferManager supports triggering flush according to the size of single ShuffleBuffer

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?
for issue https://github.com/apache/incubator-uniffle/issues/173, Add a new flush strategy for ShuffleBufferManager, flush if the size of a single buffer reaches `rss.server.buffer.flush.threshold`, which can make more full use of disk io without waiting for the buffer to reach `(rss.server.buffer.capacity * rss.server.memory.shuffle.highWaterMark.percentage)`

### How was this patch tested?
Already added

### CODE REVIEW

**Comment by leixm** [comment]
@jerqi  Can you approve running workflows for this pr plz?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 1]
```diff
@@ -284,6 +284,18 @@ public class ShuffleServerConf extends RssBaseConf {
       .defaultValue(0L)
       .withDescription("For localstorage, it will exit when the failed initialized local storage exceed the number");
 
+  public static final ConfigOption<Boolean> BUFFER_FLUSH_ENABLED = ConfigOptions
+       .key("rss.server.buffer.flush.enabled")
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:288]: Could we change the name from `rss.server.buffer.flush.enabled` to `rss.server.single.buffer.flush.enabled`?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 2]
```diff
@@ -284,6 +284,18 @@ public class ShuffleServerConf extends RssBaseConf {
...
+       .booleanType()
+       .defaultValue(false)
+       .withDescription("Whether buffer flush when size exceeded rss.server.buffer.flush.threshold");
+
+  public static final ConfigOption<Long> BUFFER_FLUSH_THRESHOLD = ConfigOptions
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:293]: Could we change the name from `rss.server.buffer.flush.threshold` to `rss.server.single.buffer.flush.threshold`?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 3]
```diff
@@ -284,6 +284,18 @@ public class ShuffleServerConf extends RssBaseConf {
...
+        .key("rss.server.buffer.flush.threshold")
+        .longType()
+        .defaultValue(32 * 1024 * 1024L)
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:296]: Could the default value be consistent with `rss.server.flush.cold.storage.threshold.size`?

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java (hunk 1 above)
Comment (leixm) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java]: done

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java (hunk 2 above)
Comment (leixm) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java]: done

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java (hunk 3 above)
Comment (leixm) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java]: done

File: README.md [hunk 4]
```diff
@@ -245,6 +245,8 @@ For more details of advanced configuration, please see [Uniffle Coordinator Guid
 |rss.storage.type|-|Supports MEMORY_LOCALFILE, MEMORY_HDFS, MEMORY_LOCALFILE_HDFS|
 |rss.server.flush.cold.storage.threshold.size|64M| The threshold of data size for LOACALFILE and HDFS if MEMORY_LOCALFILE_HDFS is used|
 |rss.server.tags|-|The comma-separated list of tags to indicate the shuffle server's attributes. It will be used as the assignment basis for the coordinator|
+|rss.server.single.buffer.flush.enabled|false|Whether single buffer flush when size exceeded rss.server.single.buffer.flush.enabled|
```
Comment (jerqi) [inline README.md:248]: Should we change 
from `exceeded rss.server.single.buffer.flush.enabled`
to `rss.server.single.buffer.flush.threshold`?

Comment (leixm) [reply README.md]: my mistake

**Review by jerqi — APPROVED** [review]
LGTM, thanks @leixm

**Comment by leixm** [comment]
Thanks for your review @jerqi

**Comment by leixm** [comment]
I think we should change `this.bufferFlushEnabled && buffer.getSize() >= this.bufferFlushThreshold` to `this.bufferFlushEnabled && buffer.getSize() > this.bufferFlushThreshold` , which should be consistent with `event.getSize() > flushColdStorageThresholdSize` @jerqi

**Comment by jerqi** [comment]
> I think we should change `this.bufferFlushEnabled && buffer.getSize() >= this.bufferFlushThreshold` to `this.bufferFlushEnabled && buffer.getSize() > this.bufferFlushThreshold` , which should be consistent with `event.getSize() > flushColdStorageThresholdSize` @jerqi

It's ok and you can raise a follow-up pr. You'd better add some comments.

**Pull request merged**